File: FurAffinity/Notifications.fs
namespace ArtworkInbox.FurAffinity

open System
open System.IO
open System.Net
open FSharp.Control.Tasks.V2
open Newtonsoft.Json

module Notifications =
    type CurrentUser = {
        name: string
        profile: string
        profile_name: string
    }
    
    type NewSubmission = {
        id: int
        title: string
        thumbnail: string
        link: string
        name: string
        profile: string
        profile_name: string
    }

    type Submissions = {
        current_user: CurrentUser
        new_submissions: NewSubmission list
    }

    let GetSubmissionsAsync fa_cookie sfw from = task {
        let req =
            sprintf "https://faexport.spangle.org.uk/notifications/submissions.json?%s&from=%d" (if sfw then "sfw=1" else "") from
            |> WebRequest.CreateHttp
        req.UserAgent <- "ArtworkInbox/0.0 (https://github.com/IsaacSchemm/ArtworkInbox)"
        req.Headers.Set("FA_COOKIE", fa_cookie)
        use! resp = req.GetResponseAsync()
        use stream = resp.GetResponseStream()
        use sr = new StreamReader(stream)
        let! json = sr.ReadToEndAsync()
        return JsonConvert.DeserializeObject<Submissions> json
    }

    type NotificationCounts = {
        submissions: int
        comments: int
        journals: int
        favorites: int
        watchers: int
        notes: int
        trouble_tickets: int
    } with
        member this.Sum =
            seq {
                this.submissions
                this.comments
                this.journals
                this.favorites
                this.watchers
                this.notes
                this.trouble_tickets
            }
            |> Seq.sum

    type NewJournal = {
        journal_id: int
        title: string
        name: string
        profile: string
        profile_name: string
        posted: string
        posted_at: DateTimeOffset
    }

    type Others = {
        current_user: CurrentUser
        notification_counts: NotificationCounts
        new_journals: NewJournal list
    }

    let GetOthersAsync fa_cookie = task {
        let req = WebRequest.CreateHttp "https://faexport.spangle.org.uk/notifications/others.json"
        req.UserAgent <- "ArtworkInbox/0.0 (https://github.com/IsaacSchemm/ArtworkInbox)"
        req.Headers.Set("FA_COOKIE", fa_cookie)
        use! resp = req.GetResponseAsync()
        use stream = resp.GetResponseStream()
        use sr = new StreamReader(stream)
        let! json = sr.ReadToEndAsync()
        return JsonConvert.DeserializeObject<Others> json
    }

File: FurAffinity/Notifications.cs
using Microsoft.FSharp.Collections;
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;

namespace ArtworkInbox.FurAffinity {
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE1006:Naming Styles")]
    public class Notifications {
        public const string FAExportHost = "faexport.spangle.org.uk";

        private readonly HttpClient _client;

        public Notifications(IHttpClientFactory httpClientFactory, string fa_cookie) {
            _client = httpClientFactory.CreateClient();
            _client.BaseAddress = new Uri($"https://{FAExportHost}");
            _client.DefaultRequestHeaders.Add("User-Agent", "ArtworkInbox/0.1 (https://github.com/IsaacSchemm/ArtworkInbox)");
            _client.DefaultRequestHeaders.Add("FA_COOKIE", fa_cookie);
        }

        public record CurrentUser(
            string name,
            string profile,
            string profile_name);

        public record NewSubmission(
            int id,
            string title,
            string thumbnail,
            string link,
            string name,
            string profile,
            string profile_name);

        public record Submissions(
            CurrentUser current_user,
            FSharpList<NewSubmission> new_submissions);

        public Task<Submissions> GetSubmissionsAsync(bool sfw, int from) =>
            _client.GetFromJsonAsync<Submissions>($"/notifications/submissions.json?{(sfw ? "sfw=1" : "")}&from={from}");

        public record NotificationCounts(
            int submissions,
            int comments,
            int journals,
            int favorites,
            int watchers,
            int notes,
            int trouble_tickets);

        public record NewJournal(
            int journal_id,
            string title,
            string name,
            string profile,
            string profile_name,
            string posted,
            DateTimeOffset posted_at);

        public record Others(
            CurrentUser current_user,
            NotificationCounts notification_counts,
            FSharpList<NewJournal> new_journals);

        public Task<Others> GetOthersAsync() =>
            _client.GetFromJsonAsync<Others>("/notifications/others.json");
    }
}
